module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,280 @@
+type ExecutionDataService struct {
+	serializer  *serializer
+	blobService network.BlobService
+	maxBlobSize int
+}
+
+func NewExecutionDataService(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+) *ExecutionDataService {
+	return &ExecutionDataService{&serializer{codec, compressor}, blobService, defaultMaxBlobSize}
+}
+
+func (s *ExecutionDataService) receiveBatch(ctx context.Context, br *BlobReceiver) ([]network.Blob, error) {
+	var blobs []network.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob network.Blob
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, err
+}
+
+func (s *ExecutionDataService) receiveBlobs(parent context.Context, br *BlobReceiver) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	var cids []cid.Cid
+
+	for {
+		batch, recvErr := s.receiveBatch(ctx, br)
+
+		for _, blob := range batch {
+			cids = append(cids, blob.Cid())
+		}
+
+		if err := s.blobService.AddBlobs(ctx, batch); err != nil {
+			return nil, err
+		}
+
+		if recvErr != nil {
+			if recvErr != ErrClosedBlobChannel {
+				// this is an unexpected error, and should never occur
+				return nil, recvErr
+			}
+
+			break
+		}
+	}
+
+	return cids, nil
+}
+
+func (s *ExecutionDataService) addBlobs(ctx context.Context, v interface{}) ([]cid.Cid, error) {
+	bcw, br := IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			return
+		}
+
+		serializeErr = bcw.Flush()
+	}()
+
+	cids, recvErr := s.receiveBlobs(ctx, br)
+
+	<-done
+
+	if recvErr != nil {
+		return nil, recvErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	cids, err := s.addBlobs(ctx, sd)
+
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	for {
+		if len(cids) == 1 {
+			return cids[0], nil
+		}
+
+		if cids, err = s.addBlobs(ctx, cids); err != nil {
+			return cid.Undef, fmt.Errorf("failed to add cid blobs: %w", err)
+		}
+	}
+}
+
+func (s *ExecutionDataService) sendBlobs(parent context.Context, bs *BlobSender, cids []cid.Cid) error {
+	defer bs.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	blobChan := s.blobService.GetBlobs(ctx, cids)
+	cachedBlobs := make(map[cid.Cid]network.Blob)
+
+	for _, c := range cids {
+		var err error
+		blob, ok := cachedBlobs[c]
+
+		if ok {
+			delete(cachedBlobs, c)
+		} else {
+			blob, err = s.findBlob(blobChan, c, cachedBlobs)
```

Comment:
could you explain in comment how the `findBlob` works along with the for loop over `cids`?

The findBlob iterates over blobChan, the questions are:
1. is the `blobChan` a closed channel? if so, please comment about it. Otherwise the operation is not safe
2. `findBlob` will iterate over the `blobChan` in each iteration, it's O(N^2), can we iterate over `blobChan` first before iterate over `cids`?

Reply:
`blobChan` is a channel of blobs that eventually gets closed when all the blobs are sent to the channel.

It is not O(N^2). When an element is iterated over in a channel, what's really happening is that the item is removed from the channel. See https://gobyexample.com/range-over-channels for an explanation of how this syntax works.

Reply:
I think the context here is that `GetBlobs` returns a channel that receives all requested blobs, but not necessarily in order so we have to read them all out and then check which ones we got.

Does `Send` require that blobs are sent in order? if it doesn't, we could flip the logic to just read from the channel and then send the blobs as they come and record which we've seen to track that it finished.

Otherwise, we'll need to block until all blobs are read. This approach will allow us to send faster than the naive blocking read approach in all but the worse case. Can you add some comments detailing the what and why

Reply:
> Does Send require that blobs are sent in order? if it doesn't, we could flip the logic to just read from the channel and then send the blobs as they come and record which we've seen to track that it finished.

The blobs must be sent in the order that they appear in the CID list.